"""
24æ™‚é–“ã‚¹ãƒˆãƒ¬ã‚¹ãƒ†ã‚¹ãƒˆã‚¹ã‚¯ãƒªãƒ—ãƒˆ
Phase 2: Pythonç¿»è¨³ã‚µãƒ¼ãƒãƒ¼ã®é•·æœŸç¨¼åƒå®‰å®šæ€§æ¤œè¨¼

ç›®çš„:
- GPU VRAMãƒªãƒ¼ã‚¯ã®æ¤œå‡º
- ãƒ¡ãƒ¢ãƒªä½¿ç”¨é‡ã®é•·æœŸæ¨ç§»æ¸¬å®š
- ã‚¯ãƒ©ãƒƒã‚·ãƒ¥å†ç¾æ€§ã®ç¢ºèª

å®Ÿè¡Œæ–¹æ³•:
    python stress_test.py --duration 24 --interval 0.5

ã‚ªãƒ—ã‚·ãƒ§ãƒ³:
    --duration: ãƒ†ã‚¹ãƒˆæ™‚é–“ï¼ˆæ™‚é–“å˜ä½ã€ãƒ‡ãƒ•ã‚©ãƒ«ãƒˆ24ï¼‰
    --interval: ãƒªã‚¯ã‚¨ã‚¹ãƒˆé–“éš”ï¼ˆç§’å˜ä½ã€ãƒ‡ãƒ•ã‚©ãƒ«ãƒˆ0.5ï¼‰
    --server-address: gRPCã‚µãƒ¼ãƒãƒ¼ã‚¢ãƒ‰ãƒ¬ã‚¹ï¼ˆãƒ‡ãƒ•ã‚©ãƒ«ãƒˆ localhost:50051ï¼‰
"""

import asyncio
import argparse
import logging
import random
import sys
from datetime import datetime, timedelta
from typing import List, Tuple

import grpc
from protos import translation_pb2, translation_pb2_grpc

# ãƒ­ã‚®ãƒ³ã‚°è¨­å®š
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    handlers=[
        logging.StreamHandler(sys.stdout),
        logging.FileHandler('stress_test.log', encoding='utf-8')
    ]
)
logger = logging.getLogger(__name__)


class StressTestRunner:
    """24æ™‚é–“ã‚¹ãƒˆãƒ¬ã‚¹ãƒ†ã‚¹ãƒˆå®Ÿè¡Œã‚¯ãƒ©ã‚¹"""

    def __init__(self, server_address: str, duration_hours: int, interval_seconds: float):
        self.server_address = server_address
        self.duration_hours = duration_hours
        self.interval_seconds = interval_seconds

        self.translation_count = 0
        self.error_count = 0
        self.start_time = None
        self.end_time = None

        # ãƒ†ã‚¹ãƒˆç”¨ç¿»è¨³ãƒ†ã‚­ã‚¹ãƒˆï¼ˆãƒãƒªã‚¨ãƒ¼ã‚·ãƒ§ãƒ³ï¼‰
        self.test_texts = [
            "ã“ã‚“ã«ã¡ã¯ã€ä¸–ç•Œï¼",
            "ä»Šæ—¥ã¯ã„ã„å¤©æ°—ã§ã™ã­ã€‚",
            "ç¿»è¨³ã‚·ã‚¹ãƒ†ãƒ ã®ãƒ†ã‚¹ãƒˆã‚’å®Ÿè¡Œã—ã¦ã„ã¾ã™ã€‚",
            "é•·æœŸç¨¼åƒã®å®‰å®šæ€§ã‚’æ¤œè¨¼ã—ã¾ã™ã€‚",
            "GPU VRAMãƒªãƒ¼ã‚¯ã‚’æ¤œå‡ºã™ã‚‹ãŸã‚ã®ã‚¹ãƒˆãƒ¬ã‚¹ãƒ†ã‚¹ãƒˆã§ã™ã€‚",
            "ãƒ¡ãƒ¢ãƒªä½¿ç”¨é‡ã®æ¨ç§»ã‚’24æ™‚é–“ç›£è¦–ã—ã¾ã™ã€‚",
            "ã‚¯ãƒ©ãƒƒã‚·ãƒ¥ãŒç™ºç”Ÿã—ãªã„ã“ã¨ã‚’ç¢ºèªã—ã¾ã™ã€‚",
            "ã“ã®ãƒ†ã‚­ã‚¹ãƒˆã¯æ—¥æœ¬èªã‹ã‚‰è‹±èªã«ç¿»è¨³ã•ã‚Œã¾ã™ã€‚",
        ]

    async def run_stress_test(self):
        """ã‚¹ãƒˆãƒ¬ã‚¹ãƒ†ã‚¹ãƒˆæœ¬ä½“"""
        logger.info("=" * 80)
        logger.info("24æ™‚é–“ã‚¹ãƒˆãƒ¬ã‚¹ãƒ†ã‚¹ãƒˆé–‹å§‹")
        logger.info("=" * 80)
        logger.info(f"ã‚µãƒ¼ãƒãƒ¼ã‚¢ãƒ‰ãƒ¬ã‚¹: {self.server_address}")
        logger.info(f"ãƒ†ã‚¹ãƒˆæ™‚é–“: {self.duration_hours} æ™‚é–“")
        logger.info(f"ãƒªã‚¯ã‚¨ã‚¹ãƒˆé–“éš”: {self.interval_seconds} ç§’")
        logger.info("=" * 80)

        self.start_time = datetime.now()
        self.end_time = self.start_time + timedelta(hours=self.duration_hours)

        logger.info(f"é–‹å§‹æ™‚åˆ»: {self.start_time}")
        logger.info(f"çµ‚äº†äºˆå®šæ™‚åˆ»: {self.end_time}")

        try:
            # gRPCãƒãƒ£ãƒãƒ«ä½œæˆ
            async with grpc.aio.insecure_channel(self.server_address) as channel:
                stub = translation_pb2_grpc.TranslationServiceStub(channel)

                # ãƒ¡ã‚¤ãƒ³ãƒ«ãƒ¼ãƒ—
                while datetime.now() < self.end_time:
                    try:
                        # ãƒ©ãƒ³ãƒ€ãƒ ã«ãƒ†ã‚¹ãƒˆãƒ†ã‚­ã‚¹ãƒˆã‚’é¸æŠ
                        test_text = random.choice(self.test_texts)

                        # ç¿»è¨³ãƒªã‚¯ã‚¨ã‚¹ãƒˆ
                        # ğŸ”¥ [PHASE2.1_FIX] Protobufå®šç¾©ã«åˆã‚ã›ã¦ãƒ•ã‚£ãƒ¼ãƒ«ãƒ‰åä¿®æ­£
                        source_language = translation_pb2.Language(code="ja")
                        target_language = translation_pb2.Language(code="en")

                        request = translation_pb2.TranslateRequest(
                            source_text=test_text,          # text â†’ source_text
                            source_language=source_language, # source_lang â†’ source_language (Languageå‹)
                            target_language=target_language  # target_lang â†’ target_language (Languageå‹)
                        )

                        # ã‚¿ã‚¤ãƒ ã‚¢ã‚¦ãƒˆè¨­å®šï¼ˆ10ç§’ï¼‰
                        response = await asyncio.wait_for(
                            stub.Translate(request),
                            timeout=10.0
                        )

                        if response.translated_text:
                            self.translation_count += 1
                        else:
                            logger.warning(f"[EMPTY_RESPONSE] ãƒªã‚¯ã‚¨ã‚¹ãƒˆ {self.translation_count + 1}")
                            self.error_count += 1

                        # çµ±è¨ˆå‡ºåŠ›ï¼ˆ100å›ã”ã¨ï¼‰
                        if self.translation_count % 100 == 0:
                            self._log_statistics()

                    except asyncio.TimeoutError:
                        self.error_count += 1
                        logger.error(f"[TIMEOUT_ERROR] ãƒªã‚¯ã‚¨ã‚¹ãƒˆ {self.translation_count + 1}")
                    except grpc.aio.AioRpcError as e:
                        self.error_count += 1
                        logger.error(
                            f"[gRPC_ERROR] Status: {e.code()}, "
                            f"Details: {e.details()}, "
                            f"ãƒªã‚¯ã‚¨ã‚¹ãƒˆ: {self.translation_count + 1}"
                        )
                    except Exception as e:
                        self.error_count += 1
                        logger.exception(f"[UNKNOWN_ERROR] {e}")

                    # ãƒªã‚¯ã‚¨ã‚¹ãƒˆé–“éš”å¾…æ©Ÿ
                    await asyncio.sleep(self.interval_seconds)

        except KeyboardInterrupt:
            logger.info("ãƒ¦ãƒ¼ã‚¶ãƒ¼ã«ã‚ˆã‚‹ä¸­æ–­")
        except Exception as e:
            logger.exception(f"è‡´å‘½çš„ã‚¨ãƒ©ãƒ¼: {e}")
        finally:
            self._log_final_statistics()

    def _log_statistics(self):
        """çµ±è¨ˆæƒ…å ±ãƒ­ã‚°å‡ºåŠ›"""
        elapsed = (datetime.now() - self.start_time).total_seconds()
        elapsed_hours = elapsed / 3600
        remaining = (self.end_time - datetime.now()).total_seconds() / 3600

        success_rate = 0.0
        if self.translation_count + self.error_count > 0:
            success_rate = (self.translation_count / (self.translation_count + self.error_count)) * 100

        req_per_min = (self.translation_count / elapsed) * 60 if elapsed > 0 else 0.0

        logger.info("=" * 80)
        logger.info("[STATISTICS]")
        logger.info(f"  æˆåŠŸç¿»è¨³æ•°: {self.translation_count}")
        logger.info(f"  ã‚¨ãƒ©ãƒ¼æ•°: {self.error_count}")
        logger.info(f"  æˆåŠŸç‡: {success_rate:.2f}%")
        logger.info(f"  çµŒéæ™‚é–“: {elapsed_hours:.2f} æ™‚é–“")
        logger.info(f"  æ®‹ã‚Šæ™‚é–“: {remaining:.2f} æ™‚é–“")
        logger.info(f"  å‡¦ç†é€Ÿåº¦: {req_per_min:.2f} req/min")
        logger.info(f"  ç¾åœ¨æ™‚åˆ»: {datetime.now()}")
        logger.info("=" * 80)

    def _log_final_statistics(self):
        """æœ€çµ‚çµ±è¨ˆæƒ…å ±ãƒ­ã‚°å‡ºåŠ›"""
        total_elapsed = (datetime.now() - self.start_time).total_seconds()
        total_hours = total_elapsed / 3600

        success_rate = 0.0
        if self.translation_count + self.error_count > 0:
            success_rate = (self.translation_count / (self.translation_count + self.error_count)) * 100

        avg_req_per_min = (self.translation_count / total_elapsed) * 60 if total_elapsed > 0 else 0.0

        logger.info("=" * 80)
        logger.info("ã‚¹ãƒˆãƒ¬ã‚¹ãƒ†ã‚¹ãƒˆçµ‚äº†")
        logger.info("=" * 80)
        logger.info(f"é–‹å§‹æ™‚åˆ»: {self.start_time}")
        logger.info(f"çµ‚äº†æ™‚åˆ»: {datetime.now()}")
        logger.info(f"ç·å®Ÿè¡Œæ™‚é–“: {total_hours:.2f} æ™‚é–“")
        logger.info(f"ç·ç¿»è¨³æ•°: {self.translation_count}")
        logger.info(f"ç·ã‚¨ãƒ©ãƒ¼æ•°: {self.error_count}")
        logger.info(f"æˆåŠŸç‡: {success_rate:.2f}%")
        logger.info(f"å¹³å‡å‡¦ç†é€Ÿåº¦: {avg_req_per_min:.2f} req/min")
        logger.info("=" * 80)


async def main():
    """ãƒ¡ã‚¤ãƒ³é–¢æ•°"""
    parser = argparse.ArgumentParser(
        description="Pythonç¿»è¨³ã‚µãƒ¼ãƒãƒ¼24æ™‚é–“ã‚¹ãƒˆãƒ¬ã‚¹ãƒ†ã‚¹ãƒˆ"
    )
    parser.add_argument(
        "--duration",
        type=int,
        default=24,
        help="ãƒ†ã‚¹ãƒˆæ™‚é–“ï¼ˆæ™‚é–“å˜ä½ã€ãƒ‡ãƒ•ã‚©ãƒ«ãƒˆ24ï¼‰"
    )
    parser.add_argument(
        "--interval",
        type=float,
        default=0.5,
        help="ãƒªã‚¯ã‚¨ã‚¹ãƒˆé–“éš”ï¼ˆç§’å˜ä½ã€ãƒ‡ãƒ•ã‚©ãƒ«ãƒˆ0.5ï¼‰"
    )
    parser.add_argument(
        "--server-address",
        type=str,
        default="localhost:50051",
        help="gRPCã‚µãƒ¼ãƒãƒ¼ã‚¢ãƒ‰ãƒ¬ã‚¹ï¼ˆãƒ‡ãƒ•ã‚©ãƒ«ãƒˆ localhost:50051ï¼‰"
    )

    args = parser.parse_args()

    # ã‚¹ãƒˆãƒ¬ã‚¹ãƒ†ã‚¹ãƒˆå®Ÿè¡Œ
    runner = StressTestRunner(
        server_address=args.server_address,
        duration_hours=args.duration,
        interval_seconds=args.interval
    )

    await runner.run_stress_test()


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        logger.info("ãƒ—ãƒ­ã‚°ãƒ©ãƒ çµ‚äº†")
    except Exception as e:
        logger.exception(f"äºˆæœŸã—ãªã„ã‚¨ãƒ©ãƒ¼: {e}")
        sys.exit(1)
